components/layout/fragment_tree/box_fragment.rs

```diff
@@ -103,6 +104,10 @@ pub(crate) struct BoxFragment {
 
     /// Additional information of from layout that could be used by Javascripts and devtools.
     pub specific_layout_info: Option<SpecificLayoutInfo>,
+
+    /// The lowest scroll tree node that would affect this fragment's border. It is used to
+    /// calculate the post composite queries (e.g., bounding box query).
+    pub scroll_tree_node_id_for_query: AtomicRefCell<Option<ScrollTreeNodeId>>,
```

Comment:
This adds a new member to every `BoxFragment` and might have a significant impact to the memory consumption. Can we avoid that?

Some of my rough thoughts are:
1. Instead of storing it in a fragment, we can traverse the scroll tree and find the node id during query
2. If that makes the performance too bad, consider caching the result somewhere else

(Thanks for @mrobinson for pointing out!)

Reply:
The issue here is that this information is lost within scroll tree construction, and we does not have a way to identify a specific fragment (except for using a strong pointer). Identifying scroll tree node id from other sources like `DOMLayoutData` might make invalidations a bit complicated. 

I guess we could put this inside `BlockLevelLayoutInfo` which only applies to block level box. My other rough thoughts is using something like `HashTable<Arc<Fragment>, ScrollTreeNodeId>` (with pointer comparison) as stacking context tree info.

Reply:
> I guess we could put this inside BlockLevelLayoutInfo which only applies to block level box.

Can we do reduce the stored information even more by storing it for scroll containers only? For other fragments, we can traverse the fragment tree first to find the first scroll container ancestor.

Using a hash map at stacking context tree info looks ugly but otherwise might be fine.

Reply:
> For each block level box, it will takes 8 bytes to store it with a pointer like `Option<Box<AtomicRefCell<ScrollTreeNodeId>>>`.

That still sounds too much... Starting with a hash map for scroll containers only sounds better, as we can trade a little bit of tidiness for performance.

Ideally this should be managed by a rare data field, but right not we don't have that on fragments.

Please also start a thread on Zulip to double check with other reviewers.